### PR TITLE
[CUSPARSE] Add CuSparseMatrixCSC * CuSparseMatrixCSC

### PR DESCRIPTION
**Pull request by amontoison**
I found today that we could use the relation `C = AB <---> Cᵀ = BᵀAᵀ` to support sparse CSC * sparse CSC products.

### CODE REVIEW

File: lib/cusparse/interfaces.jl [hunk 1]
```diff
@@ -110,6 +110,12 @@ for (taga, untaga) in tag_wrappers, (wrapa, transa, unwrapa) in op_wrappers
     end
 end
 
+for SparseMatrixType in (:CuSparseMatrixCSC, :CuSparseMatrixCSR)
+    @eval begin
+        Base.:(*)(A::$SparseMatrixType{T}, B::$SparseMatrixType{T}) where {T <: BlasFloat} = gemm('N', 'N', one(T), A, B, 'O')
```
Comment (amontoison) [inline lib/cusparse/interfaces.jl]:
```suggestion
        function Base.:(*)(A::$SparseMatrixType{T}, B::$SparseMatrixType{T}) where {T <: BlasFloat}
            CUSPARSE.version() < v"11.1.1" && throw(ErrorException("This operation is not supported by the current CUDA version."))
            gemm('N', 'N', one(T), A, B, 'O')
        end
```

File: lib/cusparse/generic.jl [hunk 2]
```diff
@@ -365,6 +365,19 @@ function gemm!(transa::SparseChar, transb::SparseChar, alpha::Number, A::CuSpars
...
 
+function gemm!(transa::SparseChar, transb::SparseChar, alpha::Number, A::CuSparseMatrixCSC{T}, B::CuSparseMatrixCSC{T},
+               beta::Number, C::CuSparseMatrixCSC{T}, index::SparseChar, algo::cusparseSpGEMMAlg_t=CUSPARSE_SPGEMM_DEFAULT) where {T}
+    # C = AB <---> Cᵀ = BᵀAᵀ
+    Aᵀ = CuSparseMatrixCSR(A.colPtr, A.rowVal, A.nzVal, reverse(size(A)))
+    Bᵀ = CuSparseMatrixCSR(B.colPtr, B.rowVal, B.nzVal, reverse(size(B)))
+    Cᵀ = CuSparseMatrixCSR(C.colPtr, C.rowVal, C.nzVal, reverse(size(C)))
+    gemm!(transb, transa, alpha, Bᵀ, Aᵀ, beta, Cᵀ, index, algo)
+    # If BᵀAᵀ and Cᵀ have the same sparsity pattern, C is already updated after the gemm! call.
+    # If BᵀAᵀ and Cᵀ doesn't the same sparsity pattern, Cᵀ was reallocated and C must be updated.
```
Comment (amontoison) [inline lib/cusparse/generic.jl]:
```suggestion
    # If BᵀAᵀ and Cᵀ don't have the same sparsity pattern, Cᵀ is reallocated and C must be updated.
```

**Comment by amontoison** [comment]
@maleadt Can you restart the failing build? It's not related to this PR.

File: test/cusparse/interfaces.jl [hunk 3]
```diff
@@ -52,6 +52,25 @@ using LinearAlgebra, SparseArrays
         end
     end
 
+    if CUSPARSE.version() >= v"11.1.1"
```
Comment (maleadt) [inline test/cusparse/interfaces.jl:56]: Add a comment why this version bound?

Comment (amontoison) [inline test/cusparse/interfaces.jl:56]:
```suggestion
    # SpGEMM was added in CUSPARSE v"11.1.1"
    if CUSPARSE.version() >= v"11.1.1"
```

**Comment by maleadt** [comment]
Thanks. CI was green before the amend (which could have used a `[skip ci]`), so let's go ahead and merge this.

**Pull request merged**